File: gtci/17-modified-binary-search/python3/1-order-agnostic-binary-search.py
class Solution:
    def search(self, arr, key):
        if not arr:
            return -1

        is_ascending = arr[0] <= arr[1]

        left = 0
        right = len(arr) - 1

        while left <= right:
            mid = (left + right) // 2

            if arr[mid] == key:
                return mid
            elif (is_ascending and arr[mid] < key) or (
                not is_ascending and arr[mid] > key
            ):
                left = mid + 1
            elif (is_ascending and arr[mid] > key) or (
                not is_ascending and arr[mid] < key
            ):
                right = mid - 1

        return -1  # element not found


if __name__ == "__main__":
    sol = Solution()

    assert sol.search([4, 6, 10], 10) == 2
    assert sol.search([1, 2, 3, 4, 5, 6, 7], 5) == 4
    assert sol.search([10, 6, 4], 10) == 0
    assert sol.search([7, 6, 5, 4, 3, 2, 1], 5) == 2
    assert sol.search([1, 3, 8, 3, 1], 3) in [1, 3]
    assert sol.search([3, 3, 3, 3, 3], 3) in [0, 1, 2, 3, 4]
    assert sol.search([], 5) == -1

    print("All test cases passed.")

File: gtci/17-modified-binary-search/cpp/1-order-agnostic-binary-search.cpp
#include <cassert>
#include <iostream>
#include <vector>

class Solution {
public:
  int search(const std::vector<int>& arr, int key) {
    if (arr.empty())
      return -1;
    
    const bool isAscending = arr[0] <= arr[arr.size() - 1];

    int left = 0;
    int right = arr.size() - 1;

    while (left <= right) {
      const int mid = left + (right - left) / 2;    // avoid potential overflow

      if (arr[mid] == key)
        return mid;
      else if ((isAscending && arr[mid] < key) || (!isAscending && arr[mid] > key))
        left = mid + 1;
      else if ((isAscending && arr[mid] > key) || (!isAscending && arr[mid] < key))
        right = mid - 1;
    }
    
    return -1;  // element not found
  }
};

int main() {
    Solution s;
    
    // ascending order
    std::vector<int> arr1 {4, 6, 10};
    int key1 = 10;
    assert(s.search(arr1, key1) == 2);

    // descending order
    std::vector<int> arr2 {10, 6, 4};
    int key2 = 10;
    assert(s.search(arr2, key2) == 0);
    
    // element not found
    std::vector<int> arr3 {10, 6, 4};
    int key3 = 5;
    assert(s.search(arr3, key3) == -1);

    std::cout << "All test cases passed." << std::endl;

    return 0;
}
